internal/dag/gatewayapi_processor.go

```diff
@@ -1761,9 +1766,43 @@ func (p *GatewayAPIProcessor) validateBackendObjectRef(backendObjectRef gatewaya
 		return nil, ref.To(resolvedRefsFalse(gatewayapi_v1beta1.RouteReasonBackendNotFound, fmt.Sprintf("service %q is invalid: %s", meta.Name, err)))
 	}
 
+	if err = validateAPPProtocol(&service.Weighted.ServicePort); err != nil {
+		return nil, ref.To(resolvedRefsFalse(gatewayapi_v1.RouteReasonUnsupportedProtocol, err.Error()))
+	}
+
 	return service, nil
 }
 
+// the ServicePort's AppProtocol must be one of the these.
+const (
+	protoK8sH2C     = "kubernetes.io/h2c"
+	protoK8sWS      = "kubernetes.io/ws"
+	protoContourH2C = "projectcontour.io/h2c"
```

Comment:
hm, I don't know if we need to redefine our own h2c protocol here since there is an upstream one available

Reply:
i define it here, just to match: httpproxy's protocol: h2c

Reply:
@skriss  WDYT?

Reply:
I don't think we need to define our own since these are standard upstream values. Let's just stick with the `kubernetes.io/*` ones.